File: src/lib/venue-search.ts
/**
 * EventFoundry Venue Search Integration
 * Provides venue search and checklist optimization for the app
 */

import { readFileSync, readdirSync } from 'fs';
import { join } from 'path';

export interface VenueEventSpace {
  space_name: string;
  min_guests: number;
  max_guests: number;
  optimal_guests?: number;
  space_type: string;
  has_stage?: boolean;
  has_dance_floor?: boolean;
  natural_lighting?: boolean;
}

export interface VenueData {
  venue_id: string;
  basic_info: {
    official_name: string;
    brand_name?: string;
    aliases: string[];
    venue_type: string;
    star_rating?: number;
    google_rating?: number;
    total_reviews: number;
  };
  location: {
    address: string;
    landmark?: string;
    district: string;
    city: string;
    state: string;
    pin_code?: string;
    coordinates?: {
      latitude: number;
      longitude: number;
    };
  };
  contact: {
    phone_primary: string;
    phone_secondary?: string;
    email?: string;
    website?: string;
    whatsapp?: string;
    booking_manager?: {
      name: string;
      designation: string;
      contact: string;
    };
  };
  capacity: {
    event_spaces: VenueEventSpace[];
    total_rooms?: number;
    parking_capacity?: number;
  };
  catering: {
    in_house_catering: boolean;
    in_house_menu_types: string[];
    outside_catering_allowed: boolean;
    bar_service_available?: boolean;
    alcohol_policy?: string;
  };
  facilities: {
    ac_available: boolean;
    backup_power: boolean;
    wifi_available: boolean;
    projector_screen: boolean;
    sound_system: boolean;
    lighting_setup: boolean;
    green_rooms: number;
    wheelchair_accessible: boolean;
    parking_type?: string;
    accommodation_available: boolean;
    accommodation_rooms: number;
  };
  pricing: {
    base_venue_charge?: Record<string, number>;
    per_plate_cost_min?: number;
    per_plate_cost_max?: number;
    security_deposit?: number;
    taxes_included: boolean;
  };
  checklist_automation: {
    auto_populate_items: string[];
    conditional_removals: string[];
    conditional_additions: string[];
  };
  search_keywords: {
    primary_keywords: string[];
    secondary_keywords: string[];
    location_keywords: string[];
  };
  data_quality_score: number;
}

export interface VenueSearchFilters {
  min_capacity?: number;
  max_capacity?: number;
  has_kitchen?: boolean;
  has_parking?: boolean;
  has_accommodation?: boolean;
  venue_type?: string;
  price_max?: number;
}

export interface VenueSearchResult extends VenueData {
  match_score: number;
  match_type: 'exact' | 'fuzzy' | 'location';
  matched_keyword?: string;
}

class VenueSearchEngine {
  private venues: VenueData[] = [];
  private venuesDir: string;

  constructor() {
    // Point to venue-crawler data directory
    this.venuesDir = join(process.cwd(), 'venue-crawler', 'data', 'venues');
    this.loadVenues();
  }

  private loadVenues(): void {
    try {
      const files = readdirSync(this.venuesDir).filter(f => f.endsWith('.json'));

      this.venues = files.map(file => {
        const content = readFileSync(join(this.venuesDir, file), 'utf-8');
        return JSON.parse(content) as VenueData;
      });

      console.log(`✅ Loaded ${this.venues.length} venues`);
    } catch (error) {
      console.warn('⚠️ Could not load venue data:', error);
      this.venues = [];
    }
  }

  /**
   * Simple string similarity scoring (Levenshtein-like)
   */
  private similarity(str1: string, str2: string): number {
    const s1 = str1.toLowerCase().trim();
    const s2 = str2.toLowerCase().trim();

    // Exact match
    if (s1 === s2) return 100;

    // Contains match
    if (s1.includes(s2) || s2.includes(s1)) return 90;

    // Token matching
    const tokens1 = s1.split(/\s+/);
    const tokens2 = s2.split(/\s+/);

    let matchedTokens = 0;
    for (const t1 of tokens1) {
      for (const t2 of tokens2) {
        if (t1 === t2 || t1.includes(t2) || t2.includes(t1)) {
          matchedTokens++;
          break;
        }
      }
    }

    const maxTokens = Math.max(tokens1.length, tokens2.length);
    return Math.round((matchedTokens / maxTokens) * 85);
  }

  /**
   * Search venues by query and filters
   */
  search(
    query: string,
    filters?: VenueSearchFilters,
    maxResults: number = 10
  ): VenueSearchResult[] {
    if (!query.trim()) return [];

    const queryLower = query.toLowerCase().trim();
    const matches: VenueSearchResult[] = [];

    for (const venue of this.venues) {
      // Search in keywords
      const allKeywords = [
        ...venue.search_keywords.primary_keywords,
        ...venue.search_keywords.secondary_keywords,
        ...venue.basic_info.aliases,
        venue.basic_info.official_name
      ];

      let bestScore = 0;
      let matchedKeyword = '';

      for (const keyword of allKeywords) {
        const score = this.similarity(queryLower, keyword);
        if (score > bestScore) {
          bestScore = score;
          matchedKeyword = keyword;
        }
      }

      // Threshold: 60% minimum
      if (bestScore >= 60) {
        matches.push({
          ...venue,
          match_score: bestScore,
          match_type: bestScore === 100 ? 'exact' : 'fuzzy',
          matched_keyword: matchedKeyword
        });
      }
    }

    // Apply filters
    let filtered = this.applyFilters(matches, filters);

    // Sort by score
    filtered.sort((a, b) => b.match_score - a.match_score);

    return filtered.slice(0, maxResults);
  }

  private applyFilters(
    venues: VenueSearchResult[],
    filters?: VenueSearchFilters
  ): VenueSearchResult[] {
    if (!filters) return venues;

    return venues.filter(venue => {
      // Capacity filter
      if (filters.min_capacity || filters.max_capacity) {
        const hasMatchingSpace = venue.capacity.event_spaces.some(space => {
          const min = filters.min_capacity || 0;
          const max = filters.max_capacity || Infinity;
          return space.max_guests >= min && space.max_guests <= max;
        });
        if (!hasMatchingSpace) return false;
      }

      // Facility filters
      if (filters.has_kitchen && !venue.catering.in_house_catering) return false;
      if (filters.has_parking && !venue.capacity.parking_capacity) return false;
      if (filters.has_accommodation && !venue.facilities.accommodation_available) return false;

      // Venue type
      if (filters.venue_type && venue.basic_info.venue_type !== filters.venue_type) return false;

      // Price filter
      if (filters.price_max && venue.pricing.per_plate_cost_max &&
          venue.pricing.per_plate_cost_max > filters.price_max) return false;

      return true;
    });
  }

  /**
   * Get venue by ID
   */
  getVenueById(venueId: string): VenueData | null {
    return this.venues.find(v => v.venue_id === venueId) || null;
  }

  /**
   * Search by location
   */
  searchByLocation(area: string, maxResults: number = 10): VenueSearchResult[] {
    const areaLower = area.toLowerCase();
    const matches: VenueSearchResult[] = [];

    for (const venue of this.venues) {
      const address = venue.location.address.toLowerCase();
      const landmark = venue.location.landmark?.toLowerCase() || '';
      const locationKeywords = venue.search_keywords.location_keywords.join(' ').toLowerCase();

      if (address.includes(areaLower) || landmark.includes(areaLower) || locationKeywords.includes(areaLower)) {
        matches.push({
          ...venue,
          match_score: 90,
          match_type: 'location'
        });
      }
    }

    return matches.slice(0, maxResults);
  }

  /**
   * Get all venues
   */
  getAllVenues(): VenueData[] {
    return this.venues;
  }

  /**
   * Get venue count
   */
  getVenueCount(): number {
    return this.venues.length;
  }
}

// Singleton instance
let searchEngineInstance: VenueSearchEngine | null = null;

export function getVenueSearchEngine(): VenueSearchEngine {
  if (!searchEngineInstance) {
    searchEngineInstance = new VenueSearchEngine();
  }
  return searchEngineInstance;
}

export default VenueSearchEngine;

File: venue-crawler/search/venue_search.py
"""
EventFoundry Venue Search Engine
Fuzzy keyword matching with filters for capacity, facilities, location
"""

import json
from pathlib import Path
from typing import List, Dict, Optional, Tuple
from fuzzywuzzy import fuzz, process
from loguru import logger

import sys
sys.path.append(str(Path(__file__).parent.parent))

from config import VENUES_DIR, FUZZY_MATCH_THRESHOLD


class VenueSearchEngine:
    """Intelligent venue search with fuzzy matching and filters"""

    def __init__(self, venues_directory: Path = VENUES_DIR):
        self.venues_dir = venues_directory
        self.venues: List[Dict] = []
        self.venue_index: Dict[str, Dict] = {}
        self.keyword_map: Dict[str, List[str]] = {}  # keyword -> [venue_ids]

        self._load_venues()
        self._build_search_index()

    def _load_venues(self):
        """Load all venue JSON files from directory"""
        logger.info(f"Loading venues from: {self.venues_dir}")

        if not self.venues_dir.exists():
            logger.warning(f"Venues directory not found: {self.venues_dir}")
            return

        venue_files = list(self.venues_dir.glob("*.json"))
        logger.info(f"Found {len(venue_files)} venue files")

        for venue_file in venue_files:
            try:
                with open(venue_file, 'r', encoding='utf-8') as f:
                    venue_data = json.load(f)
                    self.venues.append(venue_data)
                    self.venue_index[venue_data['venue_id']] = venue_data

            except Exception as e:
                logger.error(f"Error loading {venue_file}: {str(e)}")

        logger.success(f"✓ Loaded {len(self.venues)} venues")

    def _build_search_index(self):
        """Build keyword index for fast searching"""
        logger.info("Building search index...")

        for venue in self.venues:
            venue_id = venue['venue_id']
            keywords = venue.get('search_keywords', {})

            # Index primary keywords
            for keyword in keywords.get('primary_keywords', []):
                keyword_lower = keyword.lower()
                if keyword_lower not in self.keyword_map:
                    self.keyword_map[keyword_lower] = []
                self.keyword_map[keyword_lower].append(venue_id)

            # Index secondary keywords
            for keyword in keywords.get('secondary_keywords', []):
                keyword_lower = keyword.lower()
                if keyword_lower not in self.keyword_map:
                    self.keyword_map[keyword_lower] = []
                if venue_id not in self.keyword_map[keyword_lower]:
                    self.keyword_map[keyword_lower].append(venue_id)

            # Index aliases
            for alias in venue.get('basic_info', {}).get('aliases', []):
                alias_lower = alias.lower()
                if alias_lower not in self.keyword_map:
                    self.keyword_map[alias_lower] = []
                if venue_id not in self.keyword_map[alias_lower]:
                    self.keyword_map[alias_lower].append(venue_id)

        logger.success(f"✓ Indexed {len(self.keyword_map)} unique keywords")

    def search(
        self,
        query: str,
        filters: Optional[Dict] = None,
        max_results: int = 10
    ) -> List[Dict]:
        """
        Search venues with fuzzy matching and optional filters

        Args:
            query: Search query (e.g., "Casino Hotel", "venue for 300 guests")
            filters: Optional filters:
                - min_capacity: int
                - max_capacity: int
                - has_kitchen: bool
                - has_parking: bool
                - has_accommodation: bool
                - venue_type: str
                - price_max: int (per plate)
            max_results: Maximum number of results to return

        Returns:
            List of venue dictionaries with match scores
        """
        logger.info(f"Searching for: '{query}' with filters: {filters}")

        if not query.strip():
            return []

        # Step 1: Fuzzy match against all keywords
        matched_venues = self._fuzzy_match(query)

        # Step 2: Apply filters
        if filters:
            matched_venues = self._apply_filters(matched_venues, filters)

        # Step 3: Sort by match score and limit
        matched_venues.sort(key=lambda x: x['match_score'], reverse=True)
        results = matched_venues[:max_results]

        logger.success(f"✓ Found {len(results)} matching venues")
        return results

    def _fuzzy_match(self, query: str) -> List[Dict]:
        """Fuzzy match query against venue keywords"""
        query_lower = query.lower()
        matched_venues = {}

        # Exact match first
        if query_lower in self.keyword_map:
            for venue_id in self.keyword_map[query_lower]:
                venue = self.venue_index[venue_id]
                matched_venues[venue_id] = {
                    **venue,
                    'match_score': 100,
                    'match_type': 'exact'
                }
            logger.debug(f"Exact match found: {len(matched_venues)} venues")

        # Fuzzy match against all keywords
        all_keywords = list(self.keyword_map.keys())
        fuzzy_matches = process.extract(
            query_lower,
            all_keywords,
            scorer=fuzz.token_sort_ratio,
            limit=20
        )

        for matched_keyword, score in fuzzy_matches:
            if score >= FUZZY_MATCH_THRESHOLD:
                for venue_id in self.keyword_map[matched_keyword]:
                    # Don't overwrite exact matches
                    if venue_id not in matched_venues:
                        venue = self.venue_index[venue_id]
                        matched_venues[venue_id] = {
                            **venue,
                            'match_score': score,
                            'match_type': 'fuzzy',
                            'matched_keyword': matched_keyword
                        }

        logger.debug(f"Fuzzy matching found {len(matched_venues)} total venues")
        return list(matched_venues.values())

    def _apply_filters(self, venues: List[Dict], filters: Dict) -> List[Dict]:
        """Apply capacity, facility, and price filters"""
        filtered = []

        for venue in venues:
            # Capacity filter
            if 'min_capacity' in filters or 'max_capacity' in filters:
                if not self._check_capacity(venue, filters):
                    continue

            # Facility filters
            if 'has_kitchen' in filters:
                if not venue.get('catering', {}).get('in_house_catering'):
                    continue

            if 'has_parking' in filters:
                parking_capacity = venue.get('capacity', {}).get('parking_capacity')
                if not parking_capacity or parking_capacity < 1:
                    continue

            if 'has_accommodation' in filters and filters['has_accommodation']:
                if not venue.get('facilities', {}).get('accommodation_available'):
                    continue

            # Venue type filter
            if 'venue_type' in filters:
                if venue.get('basic_info', {}).get('venue_type') != filters['venue_type']:
                    continue

            # Price filter
            if 'price_max' in filters:
                price_max_venue = venue.get('pricing', {}).get('per_plate_cost_max')
                if price_max_venue and price_max_venue > filters['price_max']:
                    continue

            filtered.append(venue)

        logger.debug(f"After filtering: {len(filtered)} venues")
        return filtered

    def _check_capacity(self, venue: Dict, filters: Dict) -> bool:
        """Check if venue meets capacity requirements"""
        event_spaces = venue.get('capacity', {}).get('event_spaces', [])
        if not event_spaces:
            return False

        min_required = filters.get('min_capacity', 0)
        max_required = filters.get('max_capacity', float('inf'))

        for space in event_spaces:
            max_guests = space.get('max_guests', 0)
            if min_required <= max_guests <= max_required:
                return True

        return False

    def get_venue_by_id(self, venue_id: str) -> Optional[Dict]:
        """Get venue by exact ID"""
        return self.venue_index.get(venue_id)

    def get_all_venues(self) -> List[Dict]:
        """Get all venues"""
        return self.venues

    def get_venue_count(self) -> int:
        """Get total number of venues"""
        return len(self.venues)

    def search_by_location(self, area: str, max_results: int = 10) -> List[Dict]:
        """Search venues by location/area"""
        results = []

        for venue in self.venues:
            location = venue.get('location', {})
            address = location.get('address', '').lower()
            landmark = location.get('landmark', '').lower()

            if area.lower() in address or area.lower() in landmark:
                results.append({
                    **venue,
                    'match_score': 90,
                    'match_type': 'location'
                })

        results.sort(key=lambda x: x['match_score'], reverse=True)
        return results[:max_results]


# ============================================
# EXAMPLE USAGE & TESTING
# ============================================

if __name__ == "__main__":
    from loguru import logger
    import sys

    logger.remove()
    logger.add(sys.stderr, level="INFO")

    print("\n🔍 EventFoundry Venue Search Engine - TEST\n")

    # Initialize search engine
    search_engine = VenueSearchEngine()

    if search_engine.get_venue_count() == 0:
        print("⚠️  No venues loaded. Please run crawlers first.\n")
        sys.exit(0)

    print(f"Loaded {search_engine.get_venue_count()} venues\n")

    # Test searches
    test_queries = [
        ("Casino Hotel", {}),
        ("wedding venue for 300 guests", {"min_capacity": 250}),
        ("venue with kitchen", {"has_kitchen": True}),
        ("Marine Drive banquet", {}),
    ]

    for query, filters in test_queries:
        print(f"\n{'='*60}")
        print(f"Query: {query}")
        print(f"Filters: {filters}")
        print(f"{'='*60}")

        results = search_engine.search(query, filters, max_results=3)

        if results:
            for idx, venue in enumerate(results, 1):
                print(f"\n{idx}. {venue['basic_info']['official_name']}")
                print(f"   Match Score: {venue['match_score']}%")
                print(f"   Location: {venue['location']['address']}")

                spaces = venue['capacity']['event_spaces']
                if spaces:
                    space = spaces[0]
                    print(f"   Capacity: {space['min_guests']}-{space['max_guests']} guests")
        else:
            print("No results found.")

    print("\n" + "="*60 + "\n")
